[RFR] add admin actions

rtrox opened the pull request:
This PR adds all available admin actions to the "adminactions" extension.

API Docs for these functions are available here:
http://developer.openstack.org/api-ref-compute-v2.1.html#os-admin-actions-v2.1

The actions added are:
- createBackup
- injectNetworkInfo
- Migrate
- LiveMigrate
- ResetNetwork
- ResetState

rtrox commented:
@jrperritt I think this is ready for review, would you mind taking a look?

rtrox commented:
I should add,  these features are being implemented in preparation for a PR to rack cli to allow admin functions by Rackspace admins. Thanks for looking!

jrperritt commented:
Great, I'll take a look tomorrow.

jrperritt commented on openstack/compute/v2/extensions/adminactions/requests.go:
```diff
@@ -0,0 +1,134 @@
+	backup := make(map[string]interface{})
+
+	if opts.Name != "" {
+		backup["name"] = opts.Name
```
If `Name` is required, this should return an error if `opts.Name == ""`.

jrperritt commented on openstack/compute/v2/extensions/adminactions/requests.go:
```diff
@@ -0,0 +1,134 @@
+	if opts.Name != "" {
+		backup["name"] = opts.Name
+	}
+	if opts.BackupType != "" {
```
same thing here

jrperritt commented on openstack/compute/v2/extensions/adminactions/requests.go:
```diff
@@ -0,0 +1,134 @@
+	if opts.BackupType != "" {
+		backup["backup_type"] = opts.BackupType
+	}
+	backup["rotation"] = opts.Rotation
```
Is 0 allowed here?

rtrox replied:
Based on the unit tests upstream, it looks like 0 is an acceptable rotation:

https://github.com/openstack/nova/blob/7d4ce9e73f75ac062652a6f500924b1bcec22991/nova/tests/unit/api/openstack/compute/test_create_backup.py#L189-L212

I'll add validation to ensure non-negative integers, though.

rtrox commented:
Thanks for the feedback @jrperritt !

I've addressed the line comment concerns by adding the following validation:

```go
	if opts.Name == "" {
		return nil, fmt.Errorf("CreateBackupOpts.Name cannot be blank.")
	}
	if opts.BackupType == "" {
		return nil, fmt.Errorf("CreateBackupOpts.BackupType cannot be blank.")
	}
	if opts.Rotation < 0 {
		return nil, fmt.Errorf("CreateBackupOpts.Rotation must 0 or greater.")
	}
	backup["name"] = opts.Name
	backup["backup_type"] = opts.BackupType
	backup["rotation"] = opts.Rotation
```

As I mentioned in my line comment, the rotation parameter can be zero according to the openstack unit-tests, but I added validation to ensure a non negative value (this is also in line with upstream validation).

I believe that means I am once more ready for review, would you mind taking another look?

jrperritt commented:
LGTM. +2
